### DISCUSSION TITLE
중복 테스트 코드를 줄이기 위한 Object Mother 패턴 적용

### DISCUSSION THREAD

**Discussion by silverpoodle** [cca-ffodregamdi/running-hi-back-v2#605] [Show and Tell]
## WHY 🤷‍♀️
현재 테스트 코드에서 같은 객체를 여러 테스트 클래스에서 반복적으로 생성하고 있습니다. 
이 때문에 코드의 중복이 너무 많고 객체 생성 과정에 변경이 생기면 복잡해져요...ㅜㅜ
특히 `Member` 와 `Post` 관련된 중복코드가 넘나리 많은 것 같습니다.

## HOW 💩
`Object Mother 패턴` 은 테스트에서 자주 사용되는 객체들을 생성하는 메서드를 한 곳에 모아 `재사용` 하는 방식입니다. 
이 패턴을 사용하면 객체 생성 로직을 분리해 중복을 줄이고, 테스트 코드의 가독성도 높이고,
엔티티에 변경이 있을 때 한 번에 관리할 수 있게 됩니다!!!

```java
public final class MemberMother {

    public static Member createUserMember() {
        return Member.builder()
            .nickname("테스트멤버")
            .role(Role.USER)
            .build();
    }

    public static Member createAdminMember() {
        return Member.builder()
            .nickname("테스트 관리자")
            .role(Role.ADMIN)
            .build();
    }
}
```

```java
@BeforeEach
void setUp() {
    Member member = MemberMother.createUserMember();
    memberRepository.save(member);
    
    // 이하 생략티비
}
```

https://martinfowler.com/bliki/ObjectMother.html

리팩터링 쓴 아조씨가 제안한 방법인데요. 중복이 잦은 객체들에 적용해보는게 어떨까요?

**Comment by Subak-Uncle** [comment]
오 좋은 방법인 것 같아요!!
잘 참곳해서 리팩토링하겠습니닷!

**Comment by numerical43** [comment]
헉 안그래도 테스트 코드 만들고 수정할 때 힘들었는데 좋은 생각입니다!! 따봉!!
